### PR TITLE
feat: Marketplace get resource

### PR DESCRIPTION
**Pull request by pmarchini**

### CODE REVIEW

File: internal/cmd/marketplace/get.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+			outputFormat, err := cmd.Flags().GetString("output")
+			if err != nil {
+				return err
+			}
+
+			return getMarketplaceResource(client, id, outputFormat)
+		},
+	}
+
+	cmd.Flags().StringP("output", "o", "json", "Output format. Allowed values: json, yaml")
```
Comment (malta895) [inline internal/cmd/marketplace/get.go:60]: I would use the `JSON` and `YAML` constants also here for consistency

File: internal/cmd/marketplace/marketplace_item.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+	MountPath *string `json:"mountPath,omitempty"  yaml:"mountPath,omitempty"`
+	Name      *string `json:"name,omitempty"  yaml:"name,omitempty"`
+}
+
+type UnsecretedVariable struct {
+	NoProductionEnv string `json:"noProductionEnv"  yaml:"noProductionEnv"`
+	ProductionEnv   string `json:"productionEnv"  yaml:"productionEnv"`
+}
+
+type Type string
```
Comment (malta895) [inline internal/cmd/marketplace/marketplace_item.go:372]: Why did you call it `Type`? Isn't it too generic?

File: internal/cmd/marketplace/get.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	var marketplaceItem *Item
+	if err := resp.ParseResponse(&marketplaceItem); err != nil {
+		return nil, fmt.Errorf("error parsing response body: %w", err)
+	}
+
+	if marketplaceItem == nil {
+		return nil, fmt.Errorf("no marketplace item returned in the response")
+	}
```
Comment (malta895) [inline internal/cmd/marketplace/get.go:90]: What happens when the user enters a non existing marketplace item? Is the HTTP error returned or the latter?

In any case I would rephrase the error shown to the user to something like `marketplace item not found`

**Review by malta895 — CHANGES_REQUESTED** [review]

File: internal/cmd/marketplace/get.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		return nil, fmt.Errorf("error executing request: %w", err)
+	}
+
+	if err := resp.Error(); err != nil {
+		return nil, err
+	}
+
+	var marketplaceItem *Item
+	if err := resp.ParseResponse(&marketplaceItem); err != nil {
+		return nil, fmt.Errorf("error parsing response body: %w", err)
```
Comment (malta895) [inline internal/cmd/marketplace/get.go:85]: This error is already wrapped [here](https://github.com/mia-platform/miactl/blob/66e5605d9d731ff7a6fba07a9b74941c8e5ab8f9/internal/client/response.go#L95)

Comment (pmarchini) [reply internal/cmd/marketplace/get.go:85]: Nice catch 🚀

File: go.mod [hunk 5]
```diff
@@ -15,13 +15,27 @@ require (
 	gopkg.in/yaml.v3 v3.0.1
 )
 
+require (
```
Comment (fredmaggiowski) [inline go.mod:18]: can you run `go mod tidy`? it's weird that deps have been managed this way 🤔

File: internal/cmd/marketplace/get.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+const (
+	getMarketplaceEndpoint        = "/api/backend/marketplace/%s"
+	JSON                   string = "json"
+	YAML                   string = "yaml"
+)
+
+// GetCmd return a new cobra command for getting a single marketplace resource
+func GetCmd(options *clioptions.CLIOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "get [resource-id]",
```
Comment (fredmaggiowski) [inline internal/cmd/marketplace/get.go:36]: this id is supposed to be the _id of the item or a human readable id?

File: internal/cmd/marketplace/get.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		return nil, fmt.Errorf("missing resource id, please provide one")
+	}
+
+	resp, err := client.
+		Get().
+		APIPath(fmt.Sprintf(getMarketplaceEndpoint, resourceID)).
+		Do(context.Background())
+
+	if err != nil {
+		return nil, fmt.Errorf("error executing request: %w", err)
```
Comment (fredmaggiowski) [inline internal/cmd/marketplace/get.go:76]: this is a weird way of wrapping errors; i thinks it should be revised on the whole service; you would generally wrap received error with a constant error and not the other way around 🤔 

I'd open an issue on the matter

File: internal/cmd/marketplace/marketplace_item.go [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+
+package marketplace
+
+import (
+	"encoding/json"
+
+	"gopkg.in/yaml.v2"
+)
+
+func UnmarshalMarketplaceItem(data []byte) (Item, error) {
```
Comment (fredmaggiowski) [inline internal/cmd/marketplace/marketplace_item.go:24]: is this interface used elsewhere in the codebase? I'd expect the item to only have one marshal function that may accept the "encoding" (json/yaml) as a input rather than having a different function for each encoding.

Also I wonder if it is actually needed; aren't the standard serializer enough? can't we extend the standard instead of wrapping them?

If we still want to wrap the standard fn I'd create a single one that accept the encoding instead of having two wrappers

File: internal/cmd/marketplace/marketplace_item.go [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+type AdditionalContainerDefaultConfigMap struct {
+	Files          []PurpleFile `json:"files,omitempty"  yaml:"files,omitempty"`
+	Link           *PurpleLink  `json:"link,omitempty"  yaml:"link,omitempty"`
+	MountPath      *string      `json:"mountPath,omitempty"  yaml:"mountPath,omitempty"`
+	Name           *string      `json:"name,omitempty"  yaml:"name,omitempty"`
+	SubPaths       []string     `json:"subPaths,omitempty"  yaml:"subPaths,omitempty"`
+	ViewAsReadOnly *bool        `json:"viewAsReadOnly,omitempty"  yaml:"viewAsReadOnly,omitempty"`
+}
+
+type PurpleFile struct {
```
Comment (fredmaggiowski) [inline internal/cmd/marketplace/marketplace_item.go:179]: purple?

File: internal/cmd/marketplace/marketplace_item.go [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+type ServiceDefaultConfigMap struct {
+	Files          []FluffyFile `json:"files,omitempty"  yaml:"files,omitempty"`
+	Link           *FluffyLink  `json:"link,omitempty"  yaml:"link,omitempty"`
+	MountPath      *string      `json:"mountPath,omitempty"  yaml:"mountPath,omitempty"`
+	Name           *string      `json:"name,omitempty"  yaml:"name,omitempty"`
+	SubPaths       []string     `json:"subPaths,omitempty"  yaml:"subPaths,omitempty"`
+	ViewAsReadOnly *bool        `json:"viewAsReadOnly,omitempty"  yaml:"viewAsReadOnly,omitempty"`
+}
+
+type FluffyFile struct {
```
Comment (fredmaggiowski) [inline internal/cmd/marketplace/marketplace_item.go:288]: fluffy?

**Review by fredmaggiowski — CHANGES_REQUESTED** [review]

File: internal/cmd/marketplace/get.go (hunk 7 above)
Comment (pmarchini) [reply internal/cmd/marketplace/get.go:76]: I agree, it would be great to refactor the entire service

File: go.mod (hunk 5 above)
Comment (pmarchini) [reply go.mod:18]: Done without any effect, it's also being done by the `make lint`.
I'll try inside devcontainer

File: internal/cmd/marketplace/get.go (hunk 6 above)
Comment (pmarchini) [reply internal/cmd/marketplace/get.go:36]: the _id of the item that can be retrieved via `miactl marketplace list`

File: internal/cmd/marketplace/marketplace_item.go (hunk 9 above)
Comment (pmarchini) [reply internal/cmd/marketplace/marketplace_item.go:179]: my bad, 2 types with the same name while writing the schema

File: go.mod (hunk 5 above)
Comment (pmarchini) [reply go.mod:18]: recreated file

**Review by fredmaggiowski — APPROVED** [review]
LGTM

File: internal/encoding/encoding.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+	Indent bool
+}
+
+type UnsupportedEncodingError struct {
+	Encoding string
+}
+
+func (e UnsupportedEncodingError) Error() string {
+	return fmt.Sprintf("unsupported encoding: %s", e.Encoding)
+}
```
Comment (malta895) [inline internal/encoding/encoding.go:41]: I think it's a bit overkill to use a struct just to define an error, you could just:
```go
var ErrUnsupportedEncoding  = errrors.New("unsupported encoding")
```
and then  use it in the code: 
```go
fmt.Errorf("%w: %s", ErrUnsupportedEncoding, encoding)
```

Also I changed the error name to respect Go conventions

File: internal/encoding/encoding.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+var marshalFuncs = map[string]func(interface{}) ([]byte, error){
+	JSON: json.Marshal,
+	YAML: yaml.Marshal,
+}
+
+func UnmarshalData(data []byte, encoding string, out interface{}) error {
+	if unmarshal, ok := unmarshalFuncs[encoding]; ok {
+		return unmarshal(data, out)
+	}
+	return UnsupportedEncodingError{Encoding: encoding}
```
Comment (malta895) [inline internal/encoding/encoding.go:57]: JIf you change the error you can do this instead
```suggestion
    fmt.Errorf("%w: %s", ErrUnsupportedEncoding, encoding)
```

**Review by malta895 — CHANGES_REQUESTED** [review]

File: internal/encoding/encoding.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+	var err error
+	// Intercept panics
+	func() {
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("marshal panicked: %v", r)
+			}
+		}()
+		data, err = marshal(input)
+	}()
```
[the remark this answers is not in the document]
Comment (malta895) [reply internal/encoding/encoding.go:76]: Nevermind, I understood, is to catch the panic at line 75

File: internal/encoding/encoding.go (hunk 11 above)
Comment (pmarchini) [reply internal/encoding/encoding.go:41]: @fredmaggiowski we had a discussion about this topic.
What do you think about?

Comment (fredmaggiowski) [reply internal/encoding/encoding.go:41]: I agree with both of you, these are two different approaches that I'd use for different use cases; one is for building errors with custom data (as in this case), the other is for wrapping received errors with custom ones.

In this situation the approach taken by @pmarchini is fine for me

File: internal/encoding/encoding.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+		return nil, UnsupportedEncodingError{Encoding: encoding}
+	}
+
+	var data []byte
+	var err error
+	// Intercept panics
+	func() {
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("marshal panicked: %v", r)
```
Comment (fredmaggiowski) [inline internal/encoding/encoding.go:72]: I'd use a wrapper here for this kind of errors; this seems like a fatal error we may want to use and detect on higher level functions

File: internal/cmd/marketplace/marketplace_item.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+package marketplace
+
+import "github.com/mia-platform/miactl/internal/encoding"
+
+func Unmarshal(data []byte, encodingFormat string) (Item, error) {
+	var r Item
+
+	err := encoding.UnmarshalData(data, encodingFormat, &r)
+	if err != nil {
```
Comment (fredmaggiowski) [inline internal/cmd/marketplace/marketplace_item.go:24]:
```suggestion
	if err := encoding.UnmarshalData(data, encodingFormat, &r); err != nil {
```

File: internal/encoding/encoding.go [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+package encoding
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"gopkg.in/yaml.v2"
```
Comment (davidebianchi) [inline internal/encoding/encoding.go:24]: There is the v3

**Review by davidebianchi — CHANGES_REQUESTED** [review]

File: internal/encoding/encoding.go (hunk 16 above)
Comment (pmarchini) [reply internal/encoding/encoding.go]: done

**Review by malta895 — APPROVED** [review]

**Comment by fredmaggiowski** [comment]
@davidebianchi your comment is resolved, the PR is fine to me, if you approve the changes you asked we can move on and merge this!

**Review by davidebianchi — APPROVED** [review]
